File: game/progression.py
"""Run state, room progression, and shop upgrades.

All scaling here sits *on top of* the backend outputs. We multiply
E[0..3] by room scaling; we never replace them.
"""
from __future__ import annotations

import random
from dataclasses import dataclass, field
from typing import List, Tuple

from . import config


# Indices into pi (match flow_game.py).
ARCH_FLOW = 3


@dataclass
class RunState:
    # Progress
    room: int = 0
    kills_in_room: int = 0
    total_kills: int = 0
    # Scoring
    score: float = 0.0
    max_flow: float = 0.0
    time_survived: float = 0.0
    rooms_cleared: int = 0
    # Modifiers (shop-driven multipliers; start at identity).
    damage_mult: float = 1.0          # incoming damage scaler
    score_mult_bonus: float = 0.0     # additive to the pi-flow multiplier
    enemy_slow: float = 1.0           # <=1.0, scales enemy speed
    # Current-room bookkeeping.
    room_time_total: float = 0.0
    room_time_remaining: float = 0.0
    room_kills_required: int = 0
    is_boss_room: bool = False

    def start_next_room(self) -> None:
        self.room += 1
        self.kills_in_room = 0
        self.is_boss_room = (self.room % config.BOSS_EVERY_N_ROOMS == 0)
        if self.is_boss_room:
            self.room_time_total = 60.0 + self.room * 4.0
            self.room_kills_required = 1           # kill the boss
        else:
            self.room_time_total = (config.ROOM_BASE_TIME
                                    + self.room * config.ROOM_TIME_STEP)
            self.room_kills_required = (config.ROOM_BASE_KILLS
                                        + self.room * config.ROOM_KILL_STEP)
        self.room_time_remaining = self.room_time_total

    def clear_room(self) -> None:
        self.rooms_cleared += 1
        self.score += (config.SCORE_BOSS_CLEAR if self.is_boss_room
                       else config.SCORE_ROOM_CLEAR)

    def tick(self, pi, dt: float) -> None:
        flow = float(pi[ARCH_FLOW])
        self.max_flow = max(self.max_flow, flow)
        self.time_survived += dt
        # Time-based score: accrues faster when in Flow.
        self.score += config.SCORE_TIME_BASE * (1.0 + flow) * dt

    def score_multiplier(self, pi) -> float:
        return 1.0 + 2.0 * float(pi[ARCH_FLOW]) + self.score_mult_bonus

    def register_kill(self, pi) -> None:
        self.kills_in_room += 1
        self.total_kills += 1
        self.score += config.SCORE_KILL_BASE * self.score_multiplier(pi)

    def room_goal_met(self) -> bool:
        if self.is_boss_room:
            return False  # boss handled separately
        if self.kills_in_room >= self.room_kills_required:
            return True
        if self.room_time_remaining <= 0.0:
            return True
        return False

    def should_open_shop(self) -> bool:
        return (self.room > 0
                and self.room % config.SHOP_EVERY_N_ROOMS == 0
                and not self.is_boss_room)


# Room scaling applied to backend physics outputs.
def scale_enemy_speed(E0: float, room: int, enemy_slow: float) -> float:
    return float(E0) * (1.0 + config.ENEMY_SCALE_SPEED * room) * enemy_slow


def scale_spawn_rate(E1: float, room: int) -> float:
    return float(E1) * (1.0 + config.ENEMY_SCALE_SPAWN * room)


# ---------------------------------------------------------------------------
# Shop
# ---------------------------------------------------------------------------
UPGRADES: List[Tuple[str, str]] = [
    ("More HP",       "+20 max HP, fully heal"),
    ("Shorter Dash",  "Dash cooldown -20%"),
    ("Score Boost",   "+0.25 score multiplier"),
    ("Slow Enemies",  "Enemies move 15% slower"),
    ("Damage Armor",  "-20% damage taken"),
]


def pick_shop_choices(n: int = 3) -> List[int]:
    return random.sample(range(len(UPGRADES)), n)


def apply_upgrade(idx: int, player, run: RunState) -> str:
    name, _ = UPGRADES[idx]
    if name == "More HP":
        player.max_hp += 20
        player.hp = player.max_hp
    elif name == "Shorter Dash":
        player.dash_cd_max = max(0.25, player.dash_cd_max * 0.8)
    elif name == "Score Boost":
        run.score_mult_bonus += 0.25
    elif name == "Slow Enemies":
        run.enemy_slow *= 0.85
    elif name == "Damage Armor":
        run.damage_mult *= 0.8
    return name

File: docs/js/progression.js
/* Run state, room progression and shop upgrades — port of
 * game/progression.py.
 *
 * All scaling here sits *on top of* the backend outputs: we multiply
 * E[0..3] by room scaling, we never replace them. */

import * as C from "./config.js";

const ARCH_FLOW = C.ARCH_FLOW;

export class RunState {
  constructor() {
    // Progress
    this.room = 0;
    this.killsInRoom = 0;
    this.totalKills = 0;
    // Scoring
    this.score = 0;
    this.maxFlow = 0;
    this.timeSurvived = 0;
    this.roomsCleared = 0;
    // Modifiers (shop-driven multipliers; start at identity)
    this.damageMult = 1.0;
    this.scoreMultBonus = 0.0;
    this.enemySlow = 1.0;
    // Current-room bookkeeping
    this.roomTimeTotal = 0;
    this.roomTimeRemaining = 0;
    this.roomKillsRequired = 0;
    this.isBossRoom = false;
    this.overtime = false;
    this.overtimeRooms = 0;
    // Web-only telemetry, feeds the run report and the lifetime profile.
    this.piSeconds = [0, 0, 0, 0, 0];
    this.longestFlowStreak = 0;
    this._flowStreak = 0;
    this.upgradesTaken = [];
    this.bossesKilled = 0;
    // "timeout" | "killed" | "abandoned" — the room timer killing you at full
    // HP is the single most confusing thing this game does to a new player,
    // so the report has to be able to name the cause.
    this.deathCause = null;
  }

  startNextRoom() {
    this.room += 1;
    this.killsInRoom = 0;
    this.overtime = false;
    this.isBossRoom = this.room % C.BOSS_EVERY_N_ROOMS === 0;
    if (this.isBossRoom) {
      this.roomTimeTotal = 60.0 + this.room * 4.0;
      this.roomKillsRequired = 1; // kill the boss
    } else {
      this.roomTimeTotal = C.ROOM_BASE_TIME + this.room * C.ROOM_TIME_STEP;
      this.roomKillsRequired = C.ROOM_BASE_KILLS + this.room * C.ROOM_KILL_STEP;
    }
    this.roomTimeRemaining = this.roomTimeTotal;
  }

  clearRoom() {
    this.roomsCleared += 1;
    let base = this.isBossRoom ? C.SCORE_BOSS_CLEAR : C.SCORE_ROOM_CLEAR;
    if (this.overtime) base *= C.ROOM_OVERTIME_MULT;
    this.score += base;
    if (this.isBossRoom) this.bossesKilled += 1;
  }

  tick(pi, dt) {
    const flow = pi[ARCH_FLOW];
    this.maxFlow = Math.max(this.maxFlow, flow);
    this.timeSurvived += dt;
    // Time-based score accrues faster when in Flow.
    this.score += C.SCORE_TIME_BASE * (1.0 + flow) * dt;

    for (let k = 0; k < 5; k++) this.piSeconds[k] += pi[k] * dt;
    // "In Flow" = Flow is both dominant and confident.
    if (flow > 0.45) {
      this._flowStreak += dt;
      this.longestFlowStreak = Math.max(this.longestFlowStreak, this._flowStreak);
    } else {
      this._flowStreak = 0;
    }
  }

  scoreMultiplier(pi) {
    return 1.0 + 2.0 * pi[ARCH_FLOW] + this.scoreMultBonus;
  }

  registerKill(pi) {
    this.killsInRoom += 1;
    this.totalKills += 1;
    this.score += C.SCORE_KILL_BASE * this.scoreMultiplier(pi);
  }

  roomGoalMet() {
    if (this.isBossRoom) return false; // boss handled separately
    return this.killsInRoom >= this.roomKillsRequired;
  }

  /** Flip the room into overtime when its timer runs out.
   *
   *  Running out of time is *not* a death: the room continues with the clear
   *  bonus halved. Clamps the timer at zero so the HUD never shows a negative
   *  countdown. Returns true only on the transition tick. */
  checkOvertime() {
    if (this.isBossRoom || this.overtime) return false;
    if (this.roomTimeRemaining > 0 || this.roomGoalMet()) return false;
    this.roomTimeRemaining = 0;
    this.overtime = true;
    this.overtimeRooms += 1;
    return true;
  }

  shouldOpenShop() {
    return this.room > 0 && this.room % C.SHOP_EVERY_N_ROOMS === 0 && !this.isBossRoom;
  }

  /** Index of the archetype this run spent the most time in. */
  dominantArchetype() {
    let best = 0;
    for (let k = 1; k < 5; k++) if (this.piSeconds[k] > this.piSeconds[best]) best = k;
    return best;
  }
}

// Room scaling applied to backend physics outputs.
export function scaleEnemySpeed(E0, room, enemySlow) {
  return E0 * (1.0 + C.ENEMY_SCALE_SPEED * room) * enemySlow;
}

export function scaleSpawnRate(E1, room) {
  return E1 * (1.0 + C.ENEMY_SCALE_SPAWN * room);
}

// --- Shop ------------------------------------------------------------------

/** `favors` tags each upgrade with the archetype it plays into, purely so the
 *  shop cards can be tinted to match the state that likes them. */
export const UPGRADES = [
  { name: "More HP", desc: "+20 max HP, fully heal", favors: C.ARCH_AROUSAL, icon: "♥" },
  { name: "Shorter Dash", desc: "Dash cooldown −20%", favors: C.ARCH_FLOW, icon: "⇢" },
  { name: "Score Boost", desc: "+0.25 score multiplier", favors: C.ARCH_FLOW, icon: "★" },
  { name: "Slow Enemies", desc: "Enemies move 15% slower", favors: C.ARCH_APATHY, icon: "❄" },
  { name: "Damage Armor", desc: "−20% damage taken", favors: C.ARCH_TACTICAL, icon: "⛊" },
];

export function pickShopChoices(n = 3) {
  const pool = UPGRADES.map((_, i) => i);
  for (let i = pool.length - 1; i > 0; i--) {
    const j = (Math.random() * (i + 1)) | 0;
    [pool[i], pool[j]] = [pool[j], pool[i]];
  }
  return pool.slice(0, n);
}

export function applyUpgrade(idx, player, run) {
  const { name } = UPGRADES[idx];
  if (name === "More HP") {
    player.maxHp += 20;
    player.hp = player.maxHp;
  } else if (name === "Shorter Dash") {
    player.dashCdMax = Math.max(0.25, player.dashCdMax * 0.8);
  } else if (name === "Score Boost") {
    run.scoreMultBonus += 0.25;
  } else if (name === "Slow Enemies") {
    run.enemySlow *= 0.85;
  } else if (name === "Damage Armor") {
    run.damageMult *= 0.8;
  }
  run.upgradesTaken.push(idx);
  return name;
}
